### PR TITLE
Refactor init and run migration in projection.OnInit()

### PR DESCRIPTION
**Pull request by ysong42**

### CODE REVIEW

File: infrastructure/pg/migrate.go [hunk 1]
```diff
@@ -118,3 +118,74 @@ func (m *Migrate) Reset() error {
 
 	return nil
 }
+
+// Get the migration source files from Github, init the migrate and then run it.
+func InitAndRunMigrateFromGithub(
```
Comment (davcrypto) [inline infrastructure/pg/migrate.go:123]: Too much param, the caller side should be able to construct migrationSourceURL itself

Comment (ysong42) [reply infrastructure/pg/migrate.go:123]: How about this one?
```
func (projection *Account) OnInit() error {
	migrationSourceURL := pg.GithubMigrationSourceURL(
		appprojection.MIGRATION_GITHUB_TARGET,
		projection.config.GithubAPIUser,
		projection.config.GithubAPIToken,
		MIGRATION_DIRECOTRY,
		projection.config.MigrationRepoRef,
	)
	databaseURL := pg.MigrationDBConnString(
		projection.rdbConn.(*pg.PgxConn).ConnString(),
		MIGRATION_TABLE_NAME,
	)
	if err := pg.InitAndRunMigrate(migrationSourceURL, databaseURL); err != nil {
		projection.logger.Errorf("failed to init and run migration: %v", err)
		return err
	}

	return nil
}
```

Comment (davcrypto) [reply infrastructure/pg/migrate.go:123]: Good and change `InitAndRunMigrate ` to `migrate`/`migrateFromGitHub`

Comment (ysong42) [reply infrastructure/pg/migrate.go:123]: Sure, np~

**Review by davcrypto — APPROVED** [review]

**Review by calvinaco — APPROVED** [review]
The refactoring looks good to me. But I am thinking if we can take it further.

In particular the GitHub credential is not suppose to be changed across different projections, and the folder could possibly be simply using the projection name.

I imagined we could make the GitHub migration helper a singleton, and we can introduce a new projection base struct say RdbProjectionWithMigtatuon that always do migration based on the above rule.

On the other hand, the GitHub migration singleton can be further refactor to fulfil an interface, and we can introduce a file based migration system. Then GitHub will no longer be a dependency and give end user extra choices.

**Comment by ysong42** [comment]
> The refactoring looks good to me. But I am thinking if we can take it further.
> 
> In particular the GitHub credential is not suppose to be changed across different projections, at least for our built-in projection and the folder could possibly be simply using the projection name.
> 
> I imagined we could keep an instance of the GitHub migration helper, and we can introduce a new projection base struct say RdbProjectionWithMigtatuon that leverage the instance always do migration based on the above rule.
> 
> On the other hand, the GitHub migration helper can be further refactor to fulfil an interface, and we can introduce a file based migration system. Then GitHub will no longer be a dependency and give end user extra choices.

Thanks for the suggestion. I will try to implement it today!

File: infrastructure/pg/migrationhelper/migrationhelper.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package migrationhelper
+
+type MigrationHelper interface {
+	InitAndRunMigrate()
```
Comment (davcrypto) [inline infrastructure/pg/migrationhelper/migrationhelper.go:4]: `InitAndRunMigrate ` -> `migrate`

File: infrastructure/pg/migrationhelper/github/github.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/golang-migrate/migrate/v4"
+
+	appprojection "github.com/crypto-com/chain-indexing/projection"
+)
+
+type GithubMigrationHelper struct {
+	Config Config
+
+	MaybeSourceURL   *string
```
Comment (davcrypto) [inline infrastructure/pg/migrationhelper/github/github.go:15]: No need for pointer, we could use empty value when doing validation

File: infrastructure/pg/migrationhelper/github/github.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	MaybeDatabaseURL *string
```
Comment (davcrypto) [inline infrastructure/pg/migrationhelper/github/github.go:16]: same as above

File: usecase/projection/base.go [hunk 5]
```diff
@@ -68,3 +98,31 @@ func (base *Base) Id() string {
...
 }
+
+func setupDefaultGithubMigration(
+	projectionId string,
+	migrationHelper *github_mh.GithubMigrationHelper,
+) {
+	projectionSnakeId := strcase.ToSnake(projectionId)
+
+	if migrationHelper.MaybeSourceURL == nil {
+		migrationDirectory := fmt.Sprintf(github_mh.MIGRATION_DIRECTORY_FORMAT, projectionSnakeId)
```
Comment (davcrypto) [inline usecase/projection/base.go:109]: Can we just panic if SourceURL is not valid? Looks like there is no generic way to build a path for every projection/future entity

File: usecase/projection/base.go [hunk 6]
```diff
@@ -68,3 +98,31 @@ func (base *Base) Id() string {
 func (base *Base) Config() interface{} {
 	return base.config
 }
+
+func setupDefaultGithubMigration(
+	projectionId string,
```
Comment (davcrypto) [inline usecase/projection/base.go:103]: `projectionId` -> `id`

File: projection/bridge_activity/bridge_pending_activity/bridge_pending_activity.go [hunk 7]
```diff
@@ -41,17 +40,44 @@ type Config struct {
...
+				MaybeMigrationHelper: github_mh.NewGithubMigrationHelper(
+					github_mh.Config{
+						Config:     *config,
+						ConnString: rdbConn.(*pg.PgxConn).ConnString(),
+					},
+					migrationSourceURL,
+					"",
+				),
+			},
+		),
```
Comment (calvinaco) [inline projection/bridge_activity/bridge_pending_activity/bridge_pending_activity.go:80]: Can we pass the migration helper as an argument too? That way user can decide to use non-GitHub way to instantiate this projection.

Comment (ysong42) [reply projection/bridge_activity/bridge_pending_activity/bridge_pending_activity.go:80]: Oh, I also introduced a `FilesystemMigrationHelper`. The MaybeMigrationHelper's type is `MigrationHelper` interface.

User could switch another type of migration helper.

File: usecase/projection/base.go (hunk 6 above)
Comment (ysong42) [reply usecase/projection/base.go:103]: Will keep `projectionId` as discussed.

File: usecase/projection/base.go (hunk 5 above)
Comment (ysong42) [reply usecase/projection/base.go:109]: Will move the execute `MigrationHelper.Migrate()` logic to projection's `OnInit()`.

File: projection/bridge_activity/bridge_pending_activity/bridge_pending_activity.go (hunk 7 above)
Comment (davcrypto) [reply projection/bridge_activity/bridge_pending_activity/bridge_pending_activity.go:80]: I think we can use migration helper as an projection object argument and use it in projection.OnInit().

Comment (ysong42) [reply projection/bridge_activity/bridge_pending_activity/bridge_pending_activity.go:80]: oh, I get it now. I will put the `XXXMigrationHelper` initialization logic on the caller side. Will change it accordingly!

**Comment by ysong42** [comment]
@calvinaco @davcrypto The code is updated again!

Two major changes:
- Extract `NewXXXMigrationHelper` logic to caller side. 
- Move `Migrate()` execution to `projection.OnInit()`.

Sample code on `chain-indexing-server` side:

image

File: infrastructure/pg/migrationhelper/filesystem/filesystem.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+func NewFilesystemMigrationHelper(sourceURL, databaeURL string) *FilesystemMigrationHelper {
+	return &FilesystemMigrationHelper{
+		SourceURL:   sourceURL,
+		DatabaseURL: databaeURL,
+	}
+}
+
+// Implement MigrationHelper interface
+func (fmh *FilesystemMigrationHelper) Migrate() {
+	if fmh.SourceURL == "" {
```
Comment (davcrypto) [inline infrastructure/pg/migrationhelper/filesystem/filesystem.go:24]: Move validation to `NewFilesystemMigrationHelper`. And could we change `SourceURL` and `DatabaseURL ` to private?

File: infrastructure/pg/migrationhelper/github/github.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+) *GithubMigrationHelper {
+	return &GithubMigrationHelper{
+		SourceURL:   sourceURL,
+		DatabaseURL: databaeURL,
+	}
+}
+
+// Implement MigrationHelper interface
+func (gmh *GithubMigrationHelper) Migrate() {
+	if gmh.SourceURL == "" {
```
Comment (davcrypto) [inline infrastructure/pg/migrationhelper/github/github.go:38]: same for here

File: infrastructure/pg/migrationhelper/filesystem/filesystem.go (hunk 8 above)
Comment (ysong42) [reply infrastructure/pg/migrationhelper/filesystem/filesystem.go:24]: sure, will change it!

File: infrastructure/pg/migrationhelper/github/github.go (hunk 9 above)
Comment (ysong42) [reply infrastructure/pg/migrationhelper/github/github.go:38]: 👌

**Pull request merged**